### PR TITLE
Version 4.1

### PR DESCRIPTION
**Pull request by diginc**
Here are code changes for the new version's release.  I'm running some more automated tests locally first.

I need to check the multi-architecture version out to see if it's still suffering from a problem that appeared after it's base image changed (see #361)

### CODE REVIEW

File: Dockerfile.template [hunk 1]
```diff
@@ -34,7 +29,7 @@ EXPOSE 443
 ENV S6_LOGGING 0
 ENV S6_KEEP_ENV 1
 ENV S6_BEHAVIOUR_IF_STAGE2_FAILS 2
-ENV FTL_CMD no-daemon
+ENV FTL_CMD debug
```
Comment (diginc) [inline Dockerfile.template]: @pi-hole/api-approvers FYI no-daemon wasn't staying in the foreground for me in my hasty update last night, so I changed it to debug as a workaround to get things running (s6-service requires a foregrounded process)

File: install.sh [hunk 2]
```diff
@@ -1,22 +1,25 @@
-#!/bin/bash -ex
+#!/bin/bash -ex /
```
Comment (diginc) [inline install.sh]: wait where did this / come from.......

File: install.sh [hunk 3]
```diff
@@ -63,10 +67,12 @@ mv "${tmpLog}" /
 
 if [[ $USE_CUSTOM_BRANCHES == true ]] ; then
     ln -s /bin/true /usr/local/bin/service
-    echo "$CORE_TAG" | tee /etc/pihole/ftlbranch
+    ln -s /bin/true /usr/local/bin/update-rc.d
     echo y | bash -x pihole checkout core $CORE_TAG
     echo y | bash -x pihole checkout web $CORE_TAG
+    echo y | bash -x pihole checkout ftl ${CORE_TAG/v/}
```
Comment (diginc) [inline install.sh]: New way to checkout dev copies of FTL added as I was pre-release testing.  Noticed no 'v' was in this tag where as the others weren't so I'm search/replacing it in this bash variable.

File: s6/debian-root/etc/cont-init.d/20-start.sh [hunk 4]
```diff
@@ -10,6 +10,10 @@ fi
 # used to start dnsmasq here for gravity to use...now that conflicts port 53
 
 $bashCmd /start.sh
+# Gotta go fast, no time for gravity
```
Comment (diginc) [inline s6/debian-root/etc/cont-init.d/20-start.sh]: Maybe not required, added before I added `--dns 1.1.1.1` which actually solved my gravity issues (known issue that never manifested in tests before)

File: test/test_bash_functions.py [hunk 5]
```diff
@@ -18,7 +19,11 @@ def test_IPv6_not_True_removes_ipv6(Docker, args, expected_ipv6, expected_stdout
 
     function = Docker.run('. /bash_functions.sh ; setup_ipv4_ipv6')
     assert "Using {}".format(expected_stdout) in function.stdout
-    config = Docker.run('cat {}'.format(WEB_CONFIG)).stdout
+    if expected_stdout == 'IPv4':
+        assert 'IPv6' not in function.stdout
+    # weird slow write/sync problem; no sleep == old state of file, sleep == updated/setup state of file
+    time.sleep(1)
```
Comment (diginc) [inline test/test_bash_functions.py]: `setup_ipv4_ipv6` does what it is supposed to do according to every manual test I perform but automation/docker seems to read the file from the old state unless you wait a small amount of time at which point

File: tox.ini [hunk 6]
```diff
@@ -4,10 +4,10 @@ envlist = py27
 [testenv]
 whitelist_externals = docker
 deps = -rrequirements.txt
-# auto parallel max b/c race condition with docker fixture (I think?)
+# 2 parallel max b/c race condition with docker fixture (I think?)
```
Comment (diginc) [inline tox.ini]: My local > 2 core machine started failing more when I set this to auto.  I thought I had a parallel race condition fixed previously but there maybe more or perhaps it's related to the `time.sleep()` "fix"; slow to sync/write disks in docker's overlay2 FS.

**Comment by diginc** [comment]
Hopefully my commentary helps other understand more of these changes where I didn't leave enough code comments.

I pushed the [dev image tag](https://hub.docker.com/r/pihole/pihole/tags/) with 4.1 which needs a little more testing before going to the prod tag.

Besides changing FTL `no-daemon` to `debug` the test's break/fixes had little to do with the updating of pi-hole it's self and more to do with existing/known flakiness in tests.

File: Dockerfile.template (hunk 1 above)
Comment (Mcat12) [reply Dockerfile.template]: Thanks, I found the issue: pi-hole/FTL#430

Comment (DL6ER) [reply Dockerfile.template]: I just tried to run it as `pihole-FTL no-daemon` and - as @Mcat12 commented in his PR - observed that Ctrl+C, etc. are not caught anymore, however, `pihole-FTL` indeed stays in teh foreground and works as expected. We should continue this here: https://github.com/pi-hole/FTL/issues/432

**Comment by diginc** [comment]
The dev tag was suffering from an issue I had encountered previously with 'development' - the fix was `rm -r ~/.docker/manifests/<image>` to properly clear out that manifest's old versions.

This should be good to go, I tried the new features on my copy and test are passing-ish (albeit flakily, I have a separate plan for that).

**Comment by diginc** [comment]
As I feared the arm images still give an error when ran on real arm hardware.  The multiarch base images aren't good anymore :( I'll have to dig into this more and report back.  What I tried last time is peeking at the history of our current v4.0 image to see if I could pull out the base layer of the old version of multiarch (they don't seem to have any old tags) but the history command just says "missing" instead of giving a hash to tag.

**Comment by nemchik** [comment]
Any chance you guys might be willing to have a chat with the team at linuxserver.io to discuss the build process for this image? I asked if any of them might have any insight on this PR and they are open to discussing it on their discord server. To any who don't know lsio is a team of pretty awesome people who build a lot of pretty popular docker images.

**Comment by diginc** [comment]
Found the root of the issue : https://github.com/pi-hole/docker-pi-hole/pull/358/commits/10c44c9374e5541e0f0e5c00729444799b6e4933

I had introduced a template variable to make updating s6 a single change rather than 3 and somehow it got reverted / hardcoded again during the #358 work.

Travis is being very slow to run but tests are passing very consistently locally now so I think I'll force this through and release.

**Comment by pralor** [comment]
This pull request has been mentioned on **Pi-hole Userspace**. There might be relevant details there:

https://discourse.pi-hole.net/t/4-1-docker-logging-to-stdout/15633/5

**Pull request merged**